src/openforms/utils/files.py

```diff
@@ -0,0 +1,127 @@
+    are surpressed.
+    """
+
+    def __init__(self, filefield: FieldFile, level: LogLevel = "warning"):
+        self.filefield = filefield
+        self.level = level
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_value, exc_traceback):
+        instance = self.filefield.instance
+        if exc_value:
+            logger.log(
+                getattr(logging, self.level.upper()),
+                "File delete on model %r (pk=%s, field=%s) failed: %s",
+                type(instance),
+                instance.pk,
+                self.filefield.field.name,
+                exc_value,
+                exc_info=exc_value,
+            )
+            return True
+
+
+def _delete_obj_files(fields: List[str], obj: models.Model) -> None:
+    for name in fields:
+        filefield = getattr(obj, name)
+        with log_failed_deletes(filefield):
+            filefield.delete(save=False)
+
+
+class DeleteFileFieldFilesMixin:
+    """
+    Model mixin ensuring file deletion on database record deletion.
+
+    All file fields as part of the model will have the content removed as part of
+    the instance deletion.
+    """
+
+    def delete(self, *args, **kwargs):
+        # maybe the file deletion should be in a transaction.on_commit handler,
+        # in case the DB deletion errors but then the file is gone?
+        # Postponing that decision, as likely a number of tests will fail because they
+        # run in transactions.
+        file_field_names = get_file_field_names(type(self))
+        _delete_obj_files(file_field_names, self)
+        return super().delete(*args, **kwargs)
+
+    delete.alters_data = True
+
+
+class DeleteFilesQuerySetMixin:
+    """
+    Delete files on :meth:`django.db.models.QuerySet.delete` calls.
+
+    All file fields as part of the model will have the content removed as part of
+    the deletion of every instance in the queryset.
+
+    .. note:: For cascading deletes with related objects, you should specify
+       ``model.Meta.base_manager_name`` on the related model(s), otherwise Django
+       doesn't use your custom queryset class.
+    """
+
+    def _delete_filefields_storage(self) -> None:
+        # TODO: we might want to wrap this in transaction.on_commit, see also
+        # DeleteFileFieldFilesMixin
+        file_field_names = get_file_field_names(self.model)
+        del_query = self._chain()
+
+        # iterator in case we're dealing with large querysets and memory could be
+        # exhausted
+        for obj in del_query.iterator():
+            _delete_obj_files(file_field_names, obj)
+
+    def _raw_delete(self, using):
+        # raw delete is called in fast_deletes
+        self._delete_filefields_storage()
+        return super()._raw_delete(using)
+
+    _raw_delete.alters_data = True
+
+    def delete(self):
+        self._delete_filefields_storage()
+        return super().delete()
```

Comment:
So, yes, this should be in a `transaction.on_commit` but you commented why this might be problematic (altho non-transactional tests are possible).

Wouldn't a valid alternative be to just move the actual file deletion AFTER the record deletion? This way any record issue will prevent actuall file deletion. Something like:

```suggestion
        filefields = self._get_filefields()
        result = super().delete()
        self._delete_filefields(filefields)
        return result
```

Reply:
> on-transactional tests are possible
yes they are, but they're slow and a better alternative is the `capture_on_commit` context manager that executes them. however this PR is already large enough that for now I didn't want to bother with updating those broken tests. Maybe I should just do it though.

The order of operations was a bit of a catch 22, and I need to think about it now the context manager was added for robustness. Without that logger catching errors, you would always end up in an inconsistent state:

* records gone, but files still present (file delete errors)
* files gone, but records still present (query error)

Reply:
ok, managed to swap it around, as the capture-and-log-exceptions behaviour altered things quite a bit. I had to drop the `queryset.iterator()` because of this though, since the delete query used in the helper method comes up empty after the actual delete. I don't think it'll be a problem now in terms of memory usage, but I left a comment in case it ever becomes a problem (and then it might just be better to hook into the `pre_delete` signal).

Because of the swapping around it was easy to hook it into the `transaction.on_commit` behaviour as well, so I've made that improvement.

Reply:
correct! the context manager was added as one of the final steps, so it's behaviour was not taken into account for the order of operations yet.